### PR TITLE
withers shouldnt cause ghost bs anymore

### LINKED ISSUE #3414: Wither breaking blocks doesn't clear BlockStorage data.
### ❗ Checklist

- [X] I am using the official english version of Slimefun and did not modify the jar.
- [X] I am using an up to date "DEV" (not "RC") version of Slimefun.
- [X] I am aware that issues related to Slimefun addons need to be reported on their bug trackers and not here.
- [X] I searched for similar open issues and could not find an existing bug report on this.

### 📍 Description

Wither when breaking blocks (not the explosive projectile one or the first exploding one when spawned) doesn't clear BlockStorage data.

### 📑 Reproduction Steps

1. Spawn wither
2. Wait until the first explosion
3. Place blocks 
4. Bring the wither to that block for them to destroy.

### 💡 Expected Behavior

It does clear BlockStorage when destroyed like the explosion did.

### 📷 Screenshots / Videos

https://youtu.be/u80h35cXH3Y

### 📜 Server Log

_No response_

### 📂 `/error-reports/` folder

_No response_

### 💻 Server Software

Paper

### 🎮 Minecraft Version

1.18.x

### ⭐ Slimefun version

```
This Server uses the following setup of Slimefun:
Paper git-Paper-141 (MC: 1.18.1)
Slimefun DEV - 994 (git eac32ac9)
Metrics-Module #28
Java 17

Installed Addons: (2)
  CMILib v1.1.0.7
  CMI v9.1.0.3
```

### 🧭 Other plugins

```
Plugins (6): CMI, CMILib, Slimefun, Vault, VoidGen, WorldEdit
```

### PR DESCRIPTION
**Pull request by J3fftw1**
## Description

This fixes withers creating ghost blockstorage blocks

Leaving it at draft now till its properly tested

## Proposed changes

added a new event and a new method

## Related Issues (if applicable)

Resolves  #3414 

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.14.* - 1.19.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java [hunk 1]
```diff
@@ -50,18 +53,33 @@ public void onBlockExplode(BlockExplodeEvent e) {
...
             SlimefunItem item = BlockStorage.check(block);
 
             if (item != null) {
                 blocks.remove();
-
-                if (!(item instanceof WitherProof) && !item.callItemHandler(BlockBreakHandler.class, handler -> handleExplosion(handler, block))) {
-                    BlockStorage.clearBlockInfo(block);
-                    block.setType(Material.AIR);
-                }
+                removeResistantBlock(block);
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java:82]: This method already performs the pretty heavy lookup `BlockStorage.check`. So doing it twice FOR EVERY SINGLE block here is a huge performance drawback.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java [hunk 2]
```diff
@@ -50,18 +53,33 @@ public void onBlockExplode(BlockExplodeEvent e) {
...
 
+    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
+    public void onEntityBreak(EntityChangeBlockEvent e) {
+        if (e.getEntity().getType() == EntityType.WITHER || e.getEntity().getType() == EntityType.WITHER_SKULL) {
+            removeResistantBlock(e.getBlock());
+        }
+    }
+
+    private void removeResistantBlock(@Nonnull Block block) {
+        final SlimefunItem item = BlockStorage.check(block);
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java:64]: No need for this to be final.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java [hunk 3]
```diff
@@ -50,18 +53,33 @@ public void onBlockExplode(BlockExplodeEvent e) {
         removeResistantBlocks(e.blockList().iterator());
     }
 
+    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
+    public void onEntityBreak(EntityChangeBlockEvent e) {
+        if (e.getEntity().getType() == EntityType.WITHER || e.getEntity().getType() == EntityType.WITHER_SKULL) {
+            removeResistantBlock(e.getBlock());
+        }
+    }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java:61]: Some comments would be good, specifically referencing the issue.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by TheBusyBiscuit** [comment]
Closing this as stale

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java [hunk 4]
```diff
@@ -50,22 +53,44 @@ public void onBlockExplode(BlockExplodeEvent e) {
         removeResistantBlocks(e.blockList().iterator());
     }
 
+    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
+    public void onEntityBreak(EntityChangeBlockEvent e) {
+        if (e.getEntity().getType() == EntityType.WITHER || e.getEntity().getType() == EntityType.WITHER_SKULL) {
+
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java:59]:
```suggestion
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java [hunk 5]
```diff
@@ -50,22 +53,44 @@ public void onBlockExplode(BlockExplodeEvent e) {
...
             SlimefunItem item = BlockStorage.check(block);
 
             if (item != null) {
                 blocks.remove();
-
-                if (!(item instanceof WitherProof) && !item.callItemHandler(BlockBreakHandler.class, handler -> handleExplosion(handler, block))) {
-                    BlockStorage.clearBlockInfo(block);
-                    block.setType(Material.AIR);
-                }
+                removeResistantBlock(block);
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ExplosionsListener.java:71]: The BlockStorage check is still being made twice, you can just pass the `SlimefunItem` instance here.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by J3fftw1** [comment]
Whoops my bad this isnt stale i just completely forgot.
Will check this tomorrow.

**Comment by J3fftw1** [comment]
Could this be reviewed

**Pull request merged**